Type support for all MIDI events

VasanthManiVasi opened the pull request:
This PR adds types for the events in the midi spec and functions to encode/decode the events. The types store the decoded event data in their fields, so the values can be used directly as opposed to decoding the byte data each time it is needed. This allows code to be written concisely, improves readability and also simplifies the API.

As a preview, the function for retrieving the qpm can be written this way:
```julia
function qpm(t::MIDI.MIDIFile)
    for event in t.tracks[1].events
        if event isa SetTempo
            return 6e7 / event.tempo
        end
    end
end
```
The `spec` dictionary maps the event type byte (such as 0x51, 0x03, etc.) to the type definitions. `readmetaevent()` uses the `spec` to obtain the event type based on it's typebyte and calls the type's constructor to decode the data.
`writemetaevent()` uses `type2byte` (a reversed dictionary of `spec`) to map the event type to its type byte. 

Most MetaEvents are added as of now and the existing code is updated to work with the new event types. Please give it a try!
Any advice on the design would be very helpful! :)

Defining each type separately is quite a lot copy/paste work so it uses metaprogramming. To define new event types now, the name, the fields and the encode, decode expressions must be added to `spec`. Any advice on the readability and efficiency would also be of great help!

**Checklist**
- [x] Add Meta events
- [x] Add Midi events
- [ ] Add Sysex events
- [x] Update existing code to support the events
- [x] Update tests

VasanthManiVasi commented:
I've added types for MIDIEvents too and have updated the tests accordingly. The tests which encode order in the statusbyte (`addevent` functions) now use the channel field to encode the order.

VasanthManiVasi commented:
I think the `SysEx` event doesn't need new types since they are mainly used by manufacturers to add their ID information. Can you share your thoughts about this? @Datseris

Datseris commented:
If someone wanted to read a SYSEX event, what would they do with the new interface?

Question: what are the breaking changes that this creates?

Datseris commented:
All in all, I like the new interface. We just need to be a bit careful before merging, because I assume this includes many breaking changes. Can you perhaps do a PR at JuliaMusic_documentation in parallel to this one, so that we see how we need to update the docs to reflect these changes?

VasanthManiVasi commented:
The [`readsysexevent()`](https://github.com/JuliaMusic/MIDI.jl/blob/events/src/trackevent.jl#L147) and [`writeevent()`](https://github.com/JuliaMusic/MIDI.jl/blob/events/src/trackevent.jl#L147) can be used for SYSEX events as usual because the `SysexEvent` type is unchanged.

I think the breaking changes are that `MetaEvent` and `MIDIEvent` have become abstract types and they are supplanted by new types for each meta event and midi event. This breaks a lot of code because previously most of the code was written using `MetaEvent` and `MIDIEvent` types.

Datseris commented on src/events.jl:
```diff
@@ -0,0 +1,147 @@
+export encode, decode
+
+# Define the spec which maps from type byte to the type definitions
+const spec = Dict(
```
literal constants should be written in all capitals, `SPECS`. Furthermore, we should have a more intention revealing name. I suggest `MIDI_EVENTS_SPEC`

VasanthManiVasi replied:
`MIDI_EVENTS_SPEC` sounds great, thanks! Done in 6a5aea9.

VasanthManiVasi commented:
Sure, I will send the PR to JuliaMusic_documentation with the updated docs. Since the `readMIDIFile` and `writeMIDIFile` is deprecated, should it be replaced with `FileIO.load` and `FileIO.save` in the docs?
After a glance, it seems that `io.md` must be updated with the new types and few of the articles that use `readMIDIFile` and `writeMIDIFile` need to be updated.

Datseris commented:
oh yeah, I forgot to update the `readmidifile` stuff in the docs, please include these changes as well

Datseris commented:
What I'm thinking here is that this PR should move version to 2.0.0

Datseris commented:
@VasanthManiVasi there is one more thing you need to check: whether this PR affects performance. Can you benchmark the basic IO functions before and after this PR?

VasanthManiVasi commented:
Before the PR:
```julia
@btime load(testmidi())
970.727 μs (7965 allocations: 672.03 KiB)
```

After the PR:
```julia
@btime load(testmidi())
6.087 ms (32045 allocations: 1.74 MiB)
```

Datseris commented on src/events.jl:
```diff
@@ -0,0 +1,357 @@
+    ),
+    0x01 => :TextEvent,
+    0x02 => :CopyrightNotice,
+    0x03 => :TrackName,
```
I'm confused here. Why would the `TrackName` event have no other fields? Where is the actual name stored...?

VasanthManiVasi replied:
Since all the text events have the same field and encode/decode functions, their definitions were stored in a separate `text_defs` to avoid repetition. I've added a comment in eb63915, thanks!

Datseris commented:
Well, that's quite a massive drop in performance... 6x. I'm assuming this happens because the track events are now a type unstable array, as each event is its own type... 

I like your interface, but we should also be careful about the performance. Which, unfortunately, I don't think there is much to be done. Type instability is a no-go in Julia if we care about performance... Here's what I propose: spend some time trying to make the code faster; do some profiling, etc.. If you can't make something happen, I guess we suck it up and release 2.0 with this change. I'll also reflect a bit more and see if it is worth it. 

From my end, the concern is that this change, unfortunately, also affects the speed of loading standard notes without caring for the events/etc. 

From your end, how does this change help you? Where/how does it help you in e.g. the GSoC project?

VasanthManiVasi commented:
Since the metatype and statusbyte fields were removed from the events, there were about 3-4x more global lookups for each read/write. I've added them back and removed the lookups.

Before the PR:
```julia
@btime load(testmidi())
970.727 μs (7965 allocations: 672.03 KiB)
```
After the PR:
```julia
@btime load(testmidi())
 2.338 ms (13877 allocations: 1.02 MiB)   
```
The previous commit was ```6.087 ms (32045 allocations: 1.74 MiB)```

VasanthManiVasi commented:
This change would be of great help because it improves the speed at which work is done using the package. Especially since the data of the events are encoded/decoded automatically and since each event has it's own type, we could search for, or add or remove any event quickly at ease

Datseris commented:
2x is already good enough for me. Focus on adding the appropriate entry on changelog.md, incrementing Project.toml to 2.0.0 and opening a PR at the JuliaMusic_docs, and we're good to go.

VasanthManiVasi commented:
I've updated the docs, and #96 is fixed now.

If this is good to go, shall I be the one to merge this PR? It would be my first time ever to click that button :smile:

Datseris commented:
I just checked the documentation PR. Wait a moment. Questions:

1. Is the `textevent` function gone?
2. Are _all_ these new types exported? That is too much pollution of the name space. I would say that none of these types should be exported, but instead accessible like `MIDI.Lyric`.
3. I think it is more readable if these types end in `Event`. E.g. `LyricEvent`.

But, sure, once we have resolved the final touches, you can be the one that merges.

VasanthManiVasi commented:
> Is the textevent function gone?

The `textevent` function created the appropriate MetaEvent based on a given symbol (such as `:lyric` or `:text`). I removed it because we can create the events directly using `Lyric`, `Marker` and `TextEvent` types after this PR. We still have the `findtextevents` function for finding the text events in a midi track. 

> Are all these new types exported? That is too much pollution of the name space. I would say that none of these types should be exported, but instead accessible like MIDI.Lyric.

Got it, I've removed the exports in 3b7314c. I'll push the changes to the doc PR also since it's using the exported names.

> I think it is more readable if these types end in Event. E.g. LyricEvent. (for all events) at the moment some events end in Event, some others do not, which can lead to confusion

I agree it would be more readable that way. Done in 3b7314c.

Datseris reviewed (approved):
Alright, go ahead and merge. Do you know how to tag a new release?

VasanthManiVasi commented:
Yes, it should be done through JuliaHub right?

VasanthManiVasi commented:
Oh no, I clicked on rebase and merge.. is there a way to undo it and squash it?

Datseris commented:
Can't undo pushes on github, nope. If you want to cleanup something open a new PR :D

VasanthManiVasi commented:
Okay, I'll open a new PR if this can be cleaned up.

VasanthManiVasi commented:
I'm not sure if there's a way to fix the history through a PR. I tried some ways on a test repo but they created a new commit instead of rewriting history..  Only way I can think of is to clean this up locally and then force push to master but I don't think it's an option for us because its a public repository.  
I'll be careful when merging next time.

Datseris commented:
Oh well, what can you do. Did you tag a release? We need a release to actually test the documentation PR.

VasanthManiVasi commented:
Gotcha, I'll tag a release after #143
